sled-hardware/src/illumos/mod.rs

```diff
@@ -686,7 +701,55 @@ fn poll_device_tree(
     Ok(())
 }
 
-async fn hardware_tracking_task(
+// Using an external process, watch for the disappearance of a tofino device.
+fn monitor_tofino(log: slog::Logger, tx: broadcast::Sender<HardwareUpdate>) {
+    match std::fs::exists(TOFINO_MONITOR) {
+        Err(_) | Ok(false) => {
+            error!(&log, "tofino monitor tool not found at {}", TOFINO_MONITOR);
+            return;
+        }
+        _ => {}
+    };
+
+    loop {
+        let mut child = match std::process::Command::new(TOFINO_MONITOR).spawn()
+        {
+            Ok(c) => {
+                debug!(
+                    &log,
+                    "launched tofino monitor process as pid {}",
+                    c.id()
+                );
+                c
+            }
+            Err(e) => {
+                error!(&log, "failed to launch tofino monitor: {e:?}");
+                return;
+            }
+        };
+
+        // This child process should only exit if it has seen a tofino ASIC and
+        // then received a "tofino removed" event from the kernel.  If that
+        // happens, it should exit cleanly with an exit code of 0.  If the child
+        // exits for any other reason, we don't know what it means, and can only
+        // log the event for posterity.  In either case, we send a message to
+        // the hardware monitor task indicating that the Tofino is gone, and it
+        // will perform any necessary cleanup.  This cleanup will include
+        // shutting down the switch zone if, for some reason, it still exists.
+        match child.wait() {
+            Err(e) => error!(&log, "failed to collect exit status: {e:?}"),
+            Ok(s) => info!(&log, "child exited with code: {:?}", s.code()),
+        }
+        let _ = tx.send(HardwareUpdate::TofinoUnavailable);
+
+        // By the time we get here, the switch zone should be gone and the
+        // device tree cleaned up.  Still, it doesn't hurt to give things a
+        // moment to stabilize before restarting the monitor.
+        std::thread::sleep(std::time::Duration::from_secs(5));
```

Comment:
Could this lead to an ABA kind of thing?

Reply:
I'm not sure what that means.